src/pools/D3MAaveDaiPool.sol

```diff
@@ -0,0 +1,168 @@
+    address                  public immutable interestStrategy;
+    address                  public immutable adai; // Token representing a share of the asset pool
+
+    address public king;  // Who gets the rewards
+
+    event Collect(address indexed king, address[] assets, uint256 amt);
+
+    constructor(address hub_, address dai_, address pool_, address _rewardsClaimer) public D3MPoolBase(hub_, dai_) {
+        pool = pool_;
+
+        // Fetch the reserve data from Aave
+        (,,,,,,, address adai_, address stableDebt_, address variableDebt_, address interestStrategy_,) = LendingPoolLike(pool_).getReserveData(dai_);
+        require(adai_ != address(0), "D3MAaveDaiPool/invalid-adai");
+        require(stableDebt_ != address(0), "D3MAaveDaiPool/invalid-stableDebt");
+        require(variableDebt_ != address(0), "D3MAaveDaiPool/invalid-variableDebt");
+        require(interestStrategy_ != address(0), "D3MAaveDaiPool/invalid-interestStrategy");
+
+        adai = adai_;
+        stableDebt = ShareTokenLike(stableDebt_);
+        variableDebt = ShareTokenLike(variableDebt_);
+        interestStrategy = interestStrategy_;
+        rewardsClaimer = RewardsClaimerLike(_rewardsClaimer);
+
+        ShareTokenLike(adai_).approve(pool_, type(uint256).max);
+        TokenLike(dai_).approve(pool_, type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "D3MAaveDaiPool/overflow");
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Admin ---
+    function file(bytes32 what, address data) external auth {
+        require(live == 1, "D3MTestPool/no-file-not-live");
+
+        if (what == "king") king = data;
+        else revert("D3MPoolBase/file-unrecognized-param");
+    }
+
+    function validTarget() external view override returns (bool) {
+        (,,,,,,,,,, address strategy,) = LendingPoolLike(pool).getReserveData(address(asset));
+        return strategy == interestStrategy;
+    }
+
+    // Deposits Dai to Aave in exchange for adai which gets sent to the msg.sender
+    // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit
+    function deposit(uint256 amt) external override auth {
+        uint256 prevBalance = shareBalance();
+
+        LendingPoolLike(pool).deposit(address(asset), amt, address(this), 0);
+
+        // Verify the correct amount of shares shows up
+        require(shareBalance() == _add(prevBalance, amt), "D3MAaveDaiPool/incorrect-share-credit");
```

Comment:
This is now failing sporadically because of the off by 1 issue Sam mentioned.
I think you can change it to `>=`.

Reply:
Sorry seems I spoke too soon and the share balance can also be lower by 1 than the sum we compare it with.
Maybe the safest and most proven way is just to return to what we is running now:
https://github.com/makerdao/dss-direct-deposit/blob/master/src/DssDirectDepositAaveDai.sol#L276